core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -681,4 +642,102 @@ private void doRelease() {
             }
         }
     }
+
+    /**
+     * A variant of {@link CompletableFuture} that keeps its completion handlers into a separate list.
+     * This yields better performance than {@link CompletableFuture#handle(BiFunction)} as the number of
+     * added handlers increases because it does not create a long linked list with extra wrappers, which is
+     * especially beneficial for Java 8 which suffers a huge performance hit when complementing a future with
+     * a deep stack. See {@code cleanStack()} in Java 8 {@link CompletableFuture} for more information.
+     */
+    private final class ChannelAcquisitionFuture extends CompletableFuture<PooledChannel> {
+
+        @Nullable
+        private List<Consumer<PooledChannel>> notifications;
+
+        void addChild(SessionProtocol desiredProtocol, PoolKey key,
+                      ChannelAcquisitionFuture childPromise,
+                      ClientConnectionTimingsBuilder timingsBuilder) {
+
+            // Add to the notification list if not complete yet.
+            if (!isDone()) {
+                if (notifications == null) {
+                    notifications = new ArrayList<>();
+                }
+                notifications.add(pch -> notifyChild(desiredProtocol, key, childPromise, timingsBuilder, pch));
+                return;
+            }
+
+            // Invoke the notification task immediately if complete already.
+            notifyChild(desiredProtocol, key, childPromise, timingsBuilder,
+                        isCompletedExceptionally() ? null : getNow(null));
+        }
+
+        private void notifyChild(SessionProtocol desiredProtocol, PoolKey key,
+                                 ChannelAcquisitionFuture childPromise,
+                                 ClientConnectionTimingsBuilder timingsBuilder,
+                                 @Nullable PooledChannel pch) {
+
+            timingsBuilder.pendingAcquisitionEnd();
+
+            if (pch != null) {
+                final SessionProtocol actualProtocol = pch.protocol();
+                if (actualProtocol.isMultiplex()) {
+                    final HttpSession session = HttpSession.get(pch.get());
+                    if (session.incrementNumUnfinishedResponses()) {
+                        childPromise.complete(pch);
+                    } else if (!usePendingAcquisition(actualProtocol, key, childPromise, timingsBuilder)) {
```

Comment:
Oops, I realized that we shouln't call `timingsBuilder.pendingAcquisitionEnd();` if `usePendingAcquisition(...)` returns `true`.

Reply:
Good point. Fixed but couldn't think of a good way to test this.